sys/arm64/include/cheri.h

```diff
@@ -40,9 +40,17 @@
 #endif
 
 #ifdef _KERNEL
-#define	__USER_DDC	((void * __capability)curthread->td_frame->tf_ddc)
+#define	__USER_DDC ((cheri_getperm(__USER_PCC) & CHERI_PERM_EXECUTIVE) ? \
+    (void * __capability)curthread->td_frame->tf_ddc :			\
+    (void * __capability)curthread->td_pcb->pcb_rddc_el0)
```

Comment:
Ah, should've caught this before: the PCB is only updated lazily on context switch, so you need to do a READ_SPECIALREG_CAP

Reply:
would maybe benefit from automated regression tests as users struggle at this level of complexity?

Reply:
I don't think we even have any tests that the bounds of DDC are honoured, they're hard to test without a lot of scaffolding... this is the kind of thing you implement once and never have to touch again (once you've verified you got it right...).

Reply:
ok i will manually test to see if syscalls work for my usecase (and bonus that they now protect against accesses that are out of bounds for rddc_el0 but in bounds for ddc_el0 )

Reply:
Added to #1635